### ISSUE TITLE
vkCmdBlitImage strange behaviour since update from 07/09/2019

### ISSUE THREAD

**Issue by aerofly** [KhronosGroup/MoltenVK#658]
The last commit yesterday seems to have a broken vkCmdBlitImage implementation. When rendering to a frame buffer and then generating mipmaps from it using vkCmdBlitImage, we get slower frames each frame, e.g. the frame rate decreases over time.

I haven't looked into the latest code, but it seems like commands are piling up each frame when using vkCmdBlitImage.

**Comment by billhollings** [comment]
The latest push did include significant changes to `vkCmdBlitImage()`. 

But what you describe is a strange outcome indeed!

Can you provide a way of replicating it so I can investigate?

**Comment by billhollings** [comment]
@aerofly NM. I think I see the problem. Will push a fix ASAP.

**Comment by aerofly** [comment]
Sending a sample is difficult, but I can explain what I do and provide the function that's causing the issue:

- We render to an 8 Bit RGBA framebuffer.
- We then create the mipmap chain from this using vkCmdBlitImage
- If I comment out the line with vkCmdBlitImage performance is ok again, but I have no mipmap level

The source for this is here

```
  VkImageSubresourceRange subresourceRange{};
  subresourceRange.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;

  // transistion first mip level to transfer source
  subresourceRange.baseMipLevel   = 0;
  subresourceRange.levelCount     = 1;
  subresourceRange.baseArrayLayer = 0;
  subresourceRange.layerCount     = ArrayLayers;
  setImageLayout( vk_command_buffer, Image.GetResource(), VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, subresourceRange, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT );

  // all other levels to transfer destination
  subresourceRange.baseMipLevel   = 1;
  subresourceRange.levelCount     = MipLevels - 1;
  subresourceRange.baseArrayLayer = 0;
  subresourceRange.layerCount     = ArrayLayers;
  setImageLayout( vk_command_buffer, Image.GetResource(), VK_IMAGE_LAYOUT_UNDEFINED, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, subresourceRange, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT );

  //
  // no copy each layer down until we reach the last mipmap level
  //
  for( uint32 i = 1; i < MipLevels; ++i )
  {
    VkImageBlit imageBlit{};

    // source
    imageBlit.srcSubresource.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
    imageBlit.srcSubresource.baseArrayLayer = 0; // array_layer;
    imageBlit.srcSubresource.layerCount     = ArrayLayers;
    imageBlit.srcSubresource.mipLevel       = i - 1;
    imageBlit.srcOffsets[1].x               = int32_t( Width >> ( i - 1 ) );
    imageBlit.srcOffsets[1].y               = int32_t( Height >> ( i - 1 ) );
    imageBlit.srcOffsets[1].z               = 1;

    // destination
    imageBlit.dstSubresource.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
    imageBlit.dstSubresource.baseArrayLayer = 0; // array_layer;
    imageBlit.dstSubresource.layerCount     = ArrayLayers;
    imageBlit.dstSubresource.mipLevel       = i;
    imageBlit.dstOffsets[1].x               = int32_t( Width >> i );
    imageBlit.dstOffsets[1].y               = int32_t( Height >> i );
    imageBlit.dstOffsets[1].z               = 1;

    VkImageSubresourceRange mipSubRange{};
    mipSubRange.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
    mipSubRange.baseMipLevel   = i;
    mipSubRange.levelCount     = 1;
    mipSubRange.baseArrayLayer = 0; // array_layer;
    mipSubRange.layerCount     = ArrayLayers;

    // transiton current mip level to transfer source for read in next step
    setImageLayout( vk_command_buffer, Image.GetResource(), VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, mipSubRange, VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT );
  }

  //
  // after the loop, all mip layers are in TRANSFER_SRC layout, so transition all to SHADER_READ
  //
  subresourceRange.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
  subresourceRange.baseMipLevel   = 0;
  subresourceRange.levelCount     = MipLevels;
  subresourceRange.baseArrayLayer = 0;
  subresourceRange.layerCount     = ArrayLayers;
  setImageLayout( vk_command_buffer, Image.GetResource(), VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresourceRange, VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT );

```

**Comment by aerofly** [comment]
Ok, thank you.

**Comment by billhollings** [comment]
Fixed in PR #659. Please retest and close this issue if resolved.

**Comment by aerofly** [comment]
Thank you for the quick fix. It seems to work again. I am closing this.
